### DISCUSSION TITLE
Bitwarden_rs simply becomes unresponsive (+unhealthy) after a few seconds

### DISCUSSION THREAD

**Discussion by Chinoman10** [dani-garcia/vaultwarden#1515] [Q&A] [answered]
Hello everyone,

I was really trying to avoid opening a Discussion, or an Issue, since I always assume that if nobody is complaining about something, then it's most likely my fault, so I should do a better job at understanding what that is and fix it...
However after an entire day looking at this, I have no idea what to do anymore, so any help is greatly appreciated...

**What is happening:**
After visiting 2-4 pages, the application stops responding. In the browser, it loads forever, and in the terminal, the healthcheck just doesn't complete (so `docker ps` reports "unhealthy" status after a short while).

**Logs?**
I've enabled **trace** level logs after trying _info_ and _debug_ before, and there isn't a single line of log that helps understand WTF is going on. When the application "crashes", the logs also stop being generated, although there was one case where it still made a few lines but without any errors the same.

**Setup:**
- **OS**: Amazon Linux 2
- **Container?**: Yes, using docker as recommended
- **VM specs**: Free ec2 instance (t2.micro) from AWS (1 vCPU, 1GB RAM, 8GB SSD)
- **SSL?**: Using CloudFlare w/ Origin certificate

**Any short-term fix?:**
Yes, I back up the files, stop the container, erase (rm) it, restore the back-up, and then I run (create) it again (it's in a restart script though since I've already done this a dozen if not hundreds of times already).

I've already read a dozen issues, discussions in forums, and I'm really clueless here...
Could it be something related to the `LimitNProc` or something I saw somewhere? Or maybe something to do with a **favicon** or something?

**Comment by jjlin** [comment]
It would help to provide some info on how you started the container, what your config is, etc. Also, did you install Docker yourself or does it come pre-installed as part of the machine image?

**Reply by Chinoman10** [reply]
Hello, thanks for giving my problem a shot!
I do this command to start it:
```
  docker run -d --name ${_bw_docker_name} \
    -e ROCKET_TLS="${_bw_rocket_tls}" \
    -e LOG_FILE="${_bw_log_file}" \
    -e LOG_LEVEL="${_bw_log_level}" -e EXTENDED_LOGGING=true \
    -v ${_bw_data_folder_ext}:${_bw_data_folder_int} \
    -p ${_bw_port_ext}:${_bw_port_int} \
    bitwardenrs/server:latest
```
The variables are pre-populated by a config script that sets the variables:
```
_bw_docker_name="bitwarden"

_bw_backup_folder="${HOME}/backup-bitwarden/"

_bw_data_folder_ext="/bw-data/"
_bw_data_folder_int="/data/"

_bw_rocket_tls='{certs="/data/rsa_key.pem",key="/data/rsa_key.key"}'
_bw_log_file="${_bw_data_folder_int}bitwarden.log"
_bw_log_level="debug"  # "trace", "debug", "info", "warn", "error" or "off"

_bw_port_int=80
_bw_port_ext=443
```

What other information about the config would you want that I haven't already shared? 🤔 

As for Docker, I installed it: "`sudo yum install docker -y`". Client version is 19.03.13; API: 1.40; Server Enginer is the same version as the client.

**Comment by jjlin** [comment, marked as answer]
This sounds pretty similar to dani-garcia/bitwarden_rs#950. If exactly the same thing is happening, you should see `[WARN] Illegal SNI hostname received` messages in your logs.

But in general, it sounds like something is using up your connections. The `ROCKET_WORKERS` env var defaults to 10; if you set it to something much higher and it stays working for longer, then that would be highly indicative. Otherwise run `netstat` as described in the issue comments and see what connections are open.

**Reply by Chinoman10** [reply]
Thanks, mate, the `ROCKET_WORKERS` env var was definitely the issue here. The hostname was fine from the beginning.
May I suggest we include in the Wiki this env. var and suggest changing it to, say... 50? or 100?
Mine's using 50 connections as I'm typing this, and nobody is accessing it but me (and I've been idle for like 30m or something).
This also explains why it would work fine for 5 minutes when I was first playing with it (there were almost no web-crawlers), but after less than an hour, it wouldn't even last 1 minute without hanging.
Is there a **norobots.txt** feature we could use to prevent this from happening also? I don't want my password manager to be indexed by search engines...

**Reply by BlackDex** [reply]
It is rather weird that you need that much connections. We have 10 as the default because that should be more then enough for light used environments. And having 50 connections while not using it seems to me there is something wrong with either the reverse proxy, or you get a lot of connections somehow from somewhere.

What do the reverse proxy logs tell you? Do you see that many connections there also?

**Reply by Chinoman10** [reply]
![image](https://user-images.githubusercontent.com/8300763/111875870-44605700-899c-11eb-9a47-c8cf1bce5d03.png)
How about this...? The website became unresponsive after a few hours of me setting the env. to 300 workers... I couldn't believe I had more than 300 connections, so I checked how many there were, and _voilá_, 400+ connections lol.

I'm not using a reverse proxy... unless you're speaking about Docker?
![image](https://user-images.githubusercontent.com/8300763/111875934-8be6e300-899c-11eb-8044-f607da87f82d.png)

**Reply by jjlin** [reply]
10 workers should be plenty for most people. You should really avoid using Rocket's HTTPS in general, and especially if it's going to be publicly exposed. There are people scanning IPs continuously and they don't care about your robot preferences. For more details, see

* https://github.com/dani-garcia/bitwarden_rs/wiki/Enabling-HTTPS
* https://github.com/dani-garcia/bitwarden_rs/wiki/Hardening-Guide#https--tls-configuration

**Reply by Chinoman10** [reply]
Alright... I was trying to avoid having to set up an RP, but I guess I have no choice then.
Thanks, Jeremy you were a great help! :)

**Reply by carceneaux** [reply]
@Chinoman10 This [project](https://github.com/dadatuputi/bitwarden_gcloud/blob/master/docker-compose.yml#L36) has a reverse proxy already defined and it works great. If you're already running Docker, this is a slam dunk.

**Reply by Chinoman10** [reply]
Thank you @carceneaux, I already saw the Caddy recommendations throughout the docs.
But I like to stick to what I'm most familiar with, so I went with Nginx...
The request flows are now slightly... extensive, let's say.
User Request --> CloudFlare --> AWS ACL --> AWS Sec. Group --> AWS ec2 instance --> iptables + fail2ban --> docker --> nginx instance --> bitwarden_rs instance

Now the only thing missing is a way to ban certain IP addresses manually using fail2ban, but for that, I would have to mod Nginx so it can read the X-Forwarded-For header set by CloudFlare and pass that onto the log files so that F2B can "read them".
Anyways, this is my own problem now I guess :) I'm already quite happy that everything works pretty nicely.
